src-test/org/etools/j1939_84/controllers/part12/Part12Step01ControllerTest.java

```diff
@@ -120,10 +127,100 @@ public void testGetTotalSteps() {
+
+        runTest();
+
+        verify(engineSpeedModule, atLeastOnce()).getEngineSpeedAsString();
+        verify(engineSpeedModule, atLeastOnce()).getKeyState();
+
+        verify(mockListener).onUrgentMessage("Please turn Key ON/Engine OFF", "Adjust Key Switch", WARNING);
+
+        String expectedMessages = "Waiting for Key ON/Engine OFF..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine OFF...";
+        assertEquals(expectedMessages, listener.getMessages());
+
+        String expectedMilestones = "";
+        assertEquals(expectedMilestones, listener.getMilestones());
+
+        String expectedResults = "Initial Engine Speed = 0.0 RPMs" + NL;
+        expectedResults += "Final Engine Speed = 0.0 RPMs" + NL;
+        assertEquals(expectedResults, listener.getResults());
+    }
+
+    @Test
+    public void testWaitForKeyOff() {
+        when(engineSpeedModule.getKeyState()).thenReturn(KEY_ON_ENGINE_RUNNING,
+                                                         KEY_ON_ENGINE_RUNNING,
+                                                         KEY_ON_ENGINE_RUNNING,
+                                                         KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_ON_ENGINE_OFF);
+        when(engineSpeedModule.getEngineSpeedAsString()).thenReturn("148.6 RPMs");
+
         runTest();
 
-        assertEquals("", listener.getMessages());
-        assertEquals("", listener.getResults());
+        verify(engineSpeedModule, atLeastOnce()).getKeyState();
+        verify(engineSpeedModule, atLeastOnce()).getEngineSpeedAsString();
+
+        verify(mockListener).onUrgentMessage("Please turn Key ON/Engine OFF", "Adjust Key Switch", WARNING);
+
+        String expectedMessages = "Waiting for Key ON/Engine OFF..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine OFF..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine OFF..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine OFF..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine OFF...";
+        assertEquals(expectedMessages, listener.getMessages());
+
+        String expectedMilestones = "";
+        assertEquals(expectedMilestones, listener.getMilestones());
+
+        String expectedResults = "Initial Engine Speed = 148.6 RPMs" + NL;
+        expectedResults += "Final Engine Speed = 148.6 RPMs" + NL;
+        assertEquals(expectedResults, listener.getResults());
     }
 
+    @Test
+    public void testEngineThrowInterruptedException() {
+
+        when(engineSpeedModule.getKeyState()).thenReturn(KEY_OFF);
+        when(engineSpeedModule.getEngineSpeedAsString()).thenReturn("300.0 RPMs");
+
+        new Timer().schedule(new TimerTask() {
+            @Override
+            public void run() {
+                instance.stop();
+            }
+        }, 750);
+
```

Comment:
Again, can we convert that to an answer of NO or Cancel on the question instead of the time with the instance.stop(), please?